### PR TITLE
BUILD-186: build lastTriggeredID needs to be in status

### PR DESCRIPTION
**Pull request by gabemontero**
/assign @adambkaplan
/assign @bparees

so here is the "POC" with https://github.com/openshift/api/pull/855 's changes manually copied in, to comply with @bparees 's ask in https://github.com/openshift/api/pull/855#issuecomment-789923610

once https://github.com/openshift/api/pull/855 merges I'll remove the manual copy, do the go mod bump, and remove the WIP

### CODE REVIEW

**Comment by gabemontero** [comment]
only a sig-api-machinery flake with e2e-aws-builds, so good news there

but the e2e-aws failure `: [sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]` has to be related to our changes here, so will dive into that one ... if we are lucky, it speaks to needing both apiserver and ocm changes to make things good.

the upgrade jobs is so over the floor I'm just going to punt on that for now

**Comment by gabemontero** [comment]
> only a sig-api-machinery flake with e2e-aws-builds, so good news there
> 
> but the e2e-aws failure `: [sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]` has to be related to our changes here, so will dive into that one ... if we are lucky, it speaks to needing both apiserver and ocm changes to make things good.
> 
> the upgrade jobs is so over the floor I'm just going to punt on that for now

actually maybe it just timed out? ... the log is very weird:

```
I0304 22:44:14.301859    1095 test_context.go:459] Tolerating taints "node-role.kubernetes.io/master" when considering if nodes are ready
[BeforeEach] [Top Level]
  github.com/openshift/origin/test/extended/util/framework.go:1440
[BeforeEach] [Top Level]
  github.com/openshift/origin/test/extended/util/framework.go:1440
[BeforeEach] [Top Level]
  github.com/openshift/origin/test/extended/util/test.go:59
[BeforeEach] [sig-imageregistry][Feature:ImageTriggers] Image change build triggers
  github.com/openshift/origin/test/extended/util/client.go:140
STEP: Creating a kubernetes client
[BeforeEach] [sig-imageregistry][Feature:ImageTriggers] Image change build triggers
  github.com/openshift/origin/test/extended/util/client.go:115
Mar  4 22:44:14.783: INFO: configPath is now "/tmp/configfile766854351"
Mar  4 22:44:14.783: INFO: The user is now "e2e-test-image-change-build-trigger-9thqb-user"
Mar  4 22:44:14.783: INFO: Creating project "e2e-test-image-change-build-trigger-9thqb"
Mar  4 22:44:14.902: INFO: Waiting on permissions in project "e2e-test-image-change-build-trigger-9thqb" ...
Mar  4 22:44:14.931: INFO: Waiting for ServiceAccount "default" to be provisioned...
Mar  4 22:44:15.062: INFO: Waiting for ServiceAccount "deployer" to be provisioned...
Mar  4 22:44:15.193: INFO: Waiting for ServiceAccount "builder" to be provisioned...
Mar  4 22:44:15.322: INFO: Waiting for RoleBinding "system:image-pullers" to be provisioned...
Mar  4 22:44:16.034: INFO: Waiting for RoleBinding "system:image-builders" to be provisioned...
Mar  4 22:44:16.474: INFO: Waiting for RoleBinding "system:deployers" to be provisioned...
Mar  4 22:44:17.239: INFO: Project "e2e-test-image-change-build-trigger-9thqb" has been fully provisioned.
[It] TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]
  github.com/openshift/origin/test/extended/images/imagechange_buildtrigger.go:44
```

line 44 is basically the entry into the test ... I would have expected output as we created builds and expect failures at other line numbers if a problem did occur.

let's try again

/test e2e-aws

**Comment by gabemontero** [comment]
> > only a sig-api-machinery flake with e2e-aws-builds, so good news there
> > but the e2e-aws failure `: [sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]` has to be related to our changes here, so will dive into that one ... if we are lucky, it speaks to needing both apiserver and ocm changes to make things good.
> > the upgrade jobs is so over the floor I'm just going to punt on that for now
> 
> actually maybe it just timed out? ... the log is very weird:
> 
> ```
> I0304 22:44:14.301859    1095 test_context.go:459] Tolerating taints "node-role.kubernetes.io/master" when considering if nodes are ready
> [BeforeEach] [Top Level]
>   github.com/openshift/origin/test/extended/util/framework.go:1440
> [BeforeEach] [Top Level]
>   github.com/openshift/origin/test/extended/util/framework.go:1440
> [BeforeEach] [Top Level]
>   github.com/openshift/origin/test/extended/util/test.go:59
> [BeforeEach] [sig-imageregistry][Feature:ImageTriggers] Image change build triggers
>   github.com/openshift/origin/test/extended/util/client.go:140
> STEP: Creating a kubernetes client
> [BeforeEach] [sig-imageregistry][Feature:ImageTriggers] Image change build triggers
>   github.com/openshift/origin/test/extended/util/client.go:115
> Mar  4 22:44:14.783: INFO: configPath is now "/tmp/configfile766854351"
> Mar  4 22:44:14.783: INFO: The user is now "e2e-test-image-change-build-trigger-9thqb-user"
> Mar  4 22:44:14.783: INFO: Creating project "e2e-test-image-change-build-trigger-9thqb"
> Mar  4 22:44:14.902: INFO: Waiting on permissions in project "e2e-test-image-change-build-trigger-9thqb" ...
> Mar  4 22:44:14.931: INFO: Waiting for ServiceAccount "default" to be provisioned...
> Mar  4 22:44:15.062: INFO: Waiting for ServiceAccount "deployer" to be provisioned...
> Mar  4 22:44:15.193: INFO: Waiting for ServiceAccount "builder" to be provisioned...
> Mar  4 22:44:15.322: INFO: Waiting for RoleBinding "system:image-pullers" to be provisioned...
> Mar  4 22:44:16.034: INFO: Waiting for RoleBinding "system:image-builders" to be provisioned...
> Mar  4 22:44:16.474: INFO: Waiting for RoleBinding "system:deployers" to be provisioned...
> Mar  4 22:44:17.239: INFO: Project "e2e-test-image-change-build-trigger-9thqb" has been fully provisioned.
> [It] TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]
>   github.com/openshift/origin/test/extended/images/imagechange_buildtrigger.go:44
> ```
> 
> line 44 is basically the entry into the test ... I would have expected output as we created builds and expect failures at other line numbers if a problem did occur.
> 
> let's try again
> 
> /test e2e-aws

same weird error

think I'm going to have to run that test manually, maybe with some more prints, against my own cluster and see,
but minimally I did catch an RBAC error that OCM does not have permission to call `UpdateStatus`:

```
current.log\0606:2021-03-05T17:06:35.021706891Z E0305 17:06:35.021653       1 buildconfig_controller.go:151] buildconfigs.build.openshift.io "sti-imagestreamtag" is forbidden: User "system:serviceaccount:openshift-infra:build-config-change-controller" cannot update resource "buildconfigs/status" in API group "build.openshift.io" in the namespace "e2e-test-image-change-build-trigger-wsmnh"
```

so minimally need to sort that out if we continue to pursue https://github.com/openshift/openshift-controller-manager/pull/171/commits/205e2bd3191416e4a47dfe82b01926164b4eec02

**Comment by gabemontero** [comment]
Have https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/201 up for added permissions ^^

**Comment by gabemontero** [comment]
> only a sig-api-machinery flake with e2e-aws-builds, so good news there
> 
> but the e2e-aws failure `: [sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]` has to be related to our changes here, so will dive into that one ... if we are lucky, it speaks to needing both apiserver and ocm changes to make things good.
> 
> the upgrade jobs is so over the floor I'm just going to punt on that for now

So launching a cluster with cluster bot that has both this PR, https://github.com/openshift/openshift-apiserver/pull/191, and https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/201, the test passes:

```
started: (0/1/1) "[sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]"

passed: (7.2s) 2021-03-07T22:22:48 "[sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]"
```

that said, my cluster role binding in the OCM-O PR missed the SA to update:  `system:serviceaccount:openshift-infra:build-config-change-controller`

**Comment by gabemontero** [comment]
/hold

needs https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/201 in current form

**Comment by gabemontero** [comment]
> > only a sig-api-machinery flake with e2e-aws-builds, so good news there
> > but the e2e-aws failure `: [sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]` has to be related to our changes here, so will dive into that one ... if we are lucky, it speaks to needing both apiserver and ocm changes to make things good.
> > the upgrade jobs is so over the floor I'm just going to punt on that for now
> 
> So launching a cluster with cluster bot that has both this PR, [openshift/openshift-apiserver#191](https://github.com/openshift/openshift-apiserver/pull/191), and [openshift/cluster-openshift-controller-manager-operator#201](https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/201), the test passes:
> 
> ```
> started: (0/1/1) "[sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]"
> 
> passed: (7.2s) 2021-03-07T22:22:48 "[sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers [Suite:openshift/conformance/parallel]"
> ```
> 
> that said, my cluster role binding in the OCM-O PR missed the SA to update: `system:serviceaccount:openshift-infra:build-config-change-controller`

OK I spun up a clusterbot cluster with this PR, the OCM-O PR, but *NOT* the apiserver PR, and got this test to fail locally.

What ends up happening is the the build cannot get triggered, and but the watch employed in the image registry test never times out until the e2e's time out, hence the weird line numbers I mentioned before.

When killing the test, and leaving the test namespace up, the events generated by the OCM make the situation clear.  We are hitting an error, unrelated to permissions, updating the BC status, as noted at https://github.com/openshift/openshift-controller-manager/pull/171/files#diff-47ab93a5bb51f77551c3fbf694ec0c89b83e89d93b6d906c44398e829f0fb6c8R149 with the error string `the server could not find the requested resource`

This must be because the apiserver code which initially populates the new bc.status.ICTs fields is missing, hence the could not find error.

So bottom line, we need to merge both the OCM-O PR and the apiserver PR https://github.com/openshift/openshift-apiserver/pull/191 before this PR can merge.

**Comment by gabemontero** [comment]
host level resource issues on last e2e-aws-upgrade

**Comment by gabemontero** [comment]
https://github.com/openshift/openshift-controller-manager/pull/175/files pulled in the openshift/api changes for tracking ICTs in bc.status

will need to drop that commit from this PR

**Comment by gabemontero** [comment]
apparent unrelated failures

/retest

File: pkg/image/trigger/buildconfigs/buildconfigs.go [hunk 1]
```diff
@@ -175,9 +176,22 @@ func (r *buildConfigReactor) ImageChanged(obj runtime.Object, tagRetriever trigg
 			newSource = true
 		}
 
-		// LastTriggeredImageID is an image ref, despite the name
-		if latest == p.LastTriggeredImageID {
-			continue
+		// see if this trigger has fired previously and is recorded in status;
```
Comment (rolfedh) [inline pkg/image/trigger/buildconfigs/buildconfigs.go:179]:
```suggestion
		// Indicates whether this trigger has fired previously and is recorded in status.
```

File: pkg/image/trigger/buildconfigs/buildconfigs.go [hunk 2]
```diff
@@ -175,9 +176,22 @@ func (r *buildConfigReactor) ImageChanged(obj runtime.Object, tagRetriever trigg
...
+		// openshfit-apiserver is in charge of populating the status for image change triggers,
```
Comment (rolfedh) [inline pkg/image/trigger/buildconfigs/buildconfigs.go:180]:
```suggestion
		// openshift-apiserver updates the status of image change triggers,
```

File: pkg/image/trigger/buildconfigs/buildconfigs.go [hunk 3]
```diff
@@ -175,9 +176,22 @@ func (r *buildConfigReactor) ImageChanged(obj runtime.Object, tagRetriever trigg
...
+		// including resolving LastTriggeredImageID and LastTriggerTime
```
Comment (rolfedh) [inline pkg/image/trigger/buildconfigs/buildconfigs.go:181]:
```suggestion
		// including resolving LastTriggeredImageID and LastTriggerTime.
```

File: pkg/image/trigger/buildconfigs/buildconfigs.go [hunk 4]
```diff
@@ -175,9 +176,22 @@ func (r *buildConfigReactor) ImageChanged(obj runtime.Object, tagRetriever trigg
...
+		pStatus := ocmbuildutil.GetImageChageTriggerStatusForImageChangeTrigger(p, bc)
+		if pStatus != nil {
+			// LastTriggeredImageID is an image ref, despite the name
+			if latest == pStatus.LastTriggeredImageID {
+				continue
+			}
+		} else {
+			//TODO until openshift-apiserver is updated to maintain image triggers in the status,
```
Comment (rolfedh) [inline pkg/image/trigger/buildconfigs/buildconfigs.go:189]: Still a to-do item?

**Review by rolfedh — COMMENTED** [review]
Please see my suggestions.

File: pkg/image/trigger/buildconfigs/buildconfigs.go (hunk 4 above)
Comment (gabemontero) [reply pkg/image/trigger/buildconfigs/buildconfigs.go:189]: yeah I have a commit locally that removes this, but I was holding out on pushing it up until https://github.com/openshift/openshift-apiserver/pull/191 merged @rolfedh 

but if I'm going to pull in your suggested changes for the comments, perhaps I'll just pull the trigger on this too, and we can live with any short term e2e failures until the apiserver PR merges

File: pkg/image/trigger/buildconfigs/buildconfigs.go (hunk 1 above)
Comment (gabemontero) [reply pkg/image/trigger/buildconfigs/buildconfigs.go:179]: yeah `Indicates whether` does not quite work for me @rolfedh 

I'll go as far as `Determine whether`

**Comment by gabemontero** [comment]
> Please see my suggestions.

updates pushed @rolfedh 

ptal / apply the docs approve label at your discretion

**Comment by gabemontero** [comment]
/hold cancel

removed the short term check of both spec and status last triggered image ID

as noted previously, expect image trigger build tests to fail until https://github.com/openshift/openshift-apiserver/pull/191

just waiting on @rolfedh applying the docs approved label here, and @wewang58 to put qe approved label on both PRs, then @bparees can put the valid bz label and we start merging.

**Review by rolfedh — APPROVED** [review]

File: pkg/build/controller/buildconfig/buildconfig_controller_test.go [hunk 5]
```diff
@@ -266,11 +270,14 @@ func buildConfigWithNonZeroLastVersion() *buildv1.BuildConfig {
 	return bc
 }
 
-func buildConfigWithImageChangeTriggers(triggers []tagTriggerID) *buildv1.BuildConfig {
-	bc := baseBuildConfig()
+func buildConfigWithImageChangeTriggers(triggers []tagTriggerID, bc *buildv1.BuildConfig) *buildv1.BuildConfig {
```
Comment (coreydaley) [inline pkg/build/controller/buildconfig/buildconfig_controller_test.go:273]: If this takes a pointer, and returns a pointer, can't you eliminate the returned pointer and just modify the one that you passed into the function?

File: pkg/build/controller/buildconfig/buildconfig_controller_test.go [hunk 6]
```diff
@@ -286,9 +293,25 @@ func buildConfigWithImageChangeTriggers(triggers []tagTriggerID) *buildv1.BuildC
 	return bc
 }
 
+func buildConfigWithImageChangeTriggerStatuses(triggers []tagTriggerID, bc *buildv1.BuildConfig) *buildv1.BuildConfig {
```
Comment (coreydaley) [inline pkg/build/controller/buildconfig/buildconfig_controller_test.go:296]: If this takes a pointer, and returns a pointer, can't you eliminate the returned pointer and just modify the one that you passed into the function?

File: pkg/build/buildutil/util.go [hunk 7]
```diff
@@ -275,3 +276,29 @@ func LabelValue(name string) string {
...
+
+func fromMatch(from1 k8stypes.NamespacedName, from2 k8stypes.NamespacedName) bool {
+	if from1.Namespace == from2.Namespace && from1.Name == from2.Name {
+		return true
+	}
+	return false
+}
+
+// GetImageChageTriggerStatusForImageChangeTrigger returns the ImageChangeTrigger entry in status that corresponds
+// to the supplied ImagaeChangeTrigger entry in the spec.
```
Comment (coreydaley) [inline pkg/build/buildutil/util.go:288]: s/ImagaeChangeTrigger/ImageChangeTrigger

File: pkg/build/buildutil/util.go [hunk 8]
```diff
@@ -275,3 +276,29 @@ func LabelValue(name string) string {
 	klog.Warningf("In creating the value of the build label in the build pod, several attempts at manipulating %s to meet k8s label name requirements failed", name)
 	return name
 }
+
+func fromMatch(from1 k8stypes.NamespacedName, from2 k8stypes.NamespacedName) bool {
+	if from1.Namespace == from2.Namespace && from1.Name == from2.Name {
```
Comment (coreydaley) [inline pkg/build/buildutil/util.go:281]: This is only used once, so maybe it doesn't make sense to be it's own function, but if you keep it as it's own function, it could be simplified as `return from1.Namespace == from2.Namespace && from1.Name == from2.Name`

**Review by coreydaley — CHANGES_REQUESTED** [review]

File: pkg/image/trigger/buildconfigs/buildconfigs.go [hunk 9]
```diff
@@ -175,10 +176,20 @@ func (r *buildConfigReactor) ImageChanged(obj runtime.Object, tagRetriever trigg
 			newSource = true
 		}
 
-		// LastTriggeredImageID is an image ref, despite the name
-		if latest == p.LastTriggeredImageID {
-			continue
+		// Determine whether this trigger has fired previously and is recorded in status;
+		// openshift-apiserver updates the status of image change triggers,
+		// including resolving LastTriggeredImageID and LastTriggerTime.
+		pStatus := ocmbuildutil.GetImageChageTriggerStatusForImageChangeTrigger(p, bc)
```
Comment (rolfedh) [inline pkg/image/trigger/buildconfigs/buildconfigs.go:182]:
```suggestion
		pStatus := ocmbuildutil.GetImageChangeTriggerStatusForImageChangeTrigger(p, bc)
```
s/GetImageChageTriggerStatusForImageChangeTrigger/GetImageChangeTriggerStatusForImageChangeTrigger/ everywhere?

File: pkg/build/controller/buildconfig/buildconfig_controller_test.go (hunk 5 above)
Comment (bparees) [reply pkg/build/controller/buildconfig/buildconfig_controller_test.go:273]: no, since the pointer can be nil in this case.

(also mutating args is generally frowned upon, though this function does do it in the case that bc is not nil)

Comment (bparees) [reply pkg/build/controller/buildconfig/buildconfig_controller_test.go:273]: it's also more awkward to use a function that doesn't return a value....

result:=func(input)

vs
func(input)
result=input

Comment (gabemontero) [reply pkg/build/controller/buildconfig/buildconfig_controller_test.go:273]: things have mutated enough that I no longer need to pass in the bc param

I'll revert this to the original form like at https://github.com/openshift/openshift-controller-manager/blob/7d133c152945d0559ccf3129c81f49096277d07a/pkg/build/controller/buildconfig/buildconfig_controller_test.go#L269-L270

File: pkg/build/controller/buildconfig/buildconfig_controller_test.go (hunk 6 above)
Comment (gabemontero) [reply pkg/build/controller/buildconfig/buildconfig_controller_test.go:296]: actually at this point I think we can combine `buildConfigWithImageChangeTriggers` and `buildConfigWithImageChangeTriggersStatuses` .... I'll just do that instead

File: pkg/build/buildutil/util.go (hunk 7 above)
Comment (gabemontero) [reply pkg/build/buildutil/util.go:288]: thanks

File: pkg/image/trigger/buildconfigs/buildconfigs.go (hunk 9 above)
Comment (gabemontero) [reply pkg/image/trigger/buildconfigs/buildconfigs.go:182]: thanks

File: pkg/build/buildutil/util.go (hunk 8 above)
Comment (gabemontero) [reply pkg/build/buildutil/util.go:281]: yeah it looked klunky inlineing this with the k8styipes.NamespacedName stuff on line 299

but sure I'll switch to the return

**Comment by gabemontero** [comment]
PR updated @coreydaley @bparees @rolfedh ptal

and no lost labels :-)

**Comment by gabemontero** [comment]
the expected 

[sig-imageregistry][Feature:ImageTriggers] Image change build triggers TestMultipleImageChangeBuildTriggers

failure in both e2e-aws and e2e-aws-builds that will occur until https://github.com/openshift/openshift-apiserver/pull/191 merges

**Comment by wewang58** [comment]
Launched a cluster to test it
/lgtm

File: pkg/image/trigger/buildconfigs/buildconfigs.go [hunk 10]
```diff
@@ -175,10 +176,20 @@ func (r *buildConfigReactor) ImageChanged(obj runtime.Object, tagRetriever trigg
...
 		}
 
-		// LastTriggeredImageID is an image ref, despite the name
-		if latest == p.LastTriggeredImageID {
-			continue
+		// Determine whether this trigger has fired previously and is recorded in status;
+		// openshift-apiserver updates the status of image change triggers,
+		// including resolving LastTriggeredImageID and LastTriggerTime.
+		pStatus := ocmbuildutil.GetImageChangeTriggerStatusForImageChangeTrigger(p, bc)
+		if pStatus != nil {
```
Comment (coreydaley) [inline pkg/image/trigger/buildconfigs/buildconfigs.go:183]: nit:
`
if pStatus != nil && latest == pStatus.LastTriggeredImageID {
...
}
`

File: pkg/build/controller/buildconfig/buildconfig_controller_test.go [hunk 11]
```diff
@@ -282,13 +283,23 @@ func buildConfigWithImageChangeTriggers(triggers []tagTriggerID) *buildv1.BuildC
 			Type:        buildv1.ImageChangeBuildTriggerType,
 			ImageChange: imageChangeTrigger,
 		})
+		imageChangeTriggerStatus := buildv1.ImageChangeTriggerStatus{
+			LastTriggeredImageID: trigger.LastTriggeredId,
+		}
+		if len(trigger.ImageStreamTag) > 0 {
```
Comment (coreydaley) [inline pkg/build/controller/buildconfig/buildconfig_controller_test.go:289]: nit:
`if len(...) != 0 {`

**Review by coreydaley — APPROVED** [review]
just a couple of small nits, but otherwise /lgtm

**Review by adambkaplan — APPROVED** [review]
/approve

**Comment by adambkaplan** [comment]
note - won't merge without valid-bug, nit fixes will undo lgtm

File: pkg/build/buildutil/util.go [hunk 12]
```diff
@@ -275,3 +275,48 @@ func LabelValue(name string) string {
...
+	}
+	if from1.Kind == from2.Kind && from1.Namespace == from2.Namespace && from1.Name == from2.Name {
+		return true
+	}
+	return false
+}
+
+// GetImageChageTriggerStatusForImageChangeTrigger returns the ImageChangeTrigger entry in status that corresponds
+// to the supplied ImagaeChangeTrigger entry in the spec.  The index into the array is also returned in case modifications are needed.
+func GetImageChageTriggerStatusForImageChangeTrigger(triggerSpec *buildv1.ImageChangeTrigger, bc *buildv1.BuildConfig) (*buildv1.ImageChangeTriggerStatus, int) {
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/build/buildutil/util.go]: the multi return is outdated @adambkaplan  ... that function only returns `*buildv1.ImageChangeTriggerStatus` now

File: pkg/image/controller/trigger/image_trigger_controller_test.go [hunk 13]
```diff
@@ -643,7 +648,6 @@ func scenario_1_buildConfig_imageSource_defaultImageTag() *buildv1.BuildConfig {
 						Namespace: "other"},
 					LastTriggeredImageID: "image/result:2",
 				}},
-				{ImageChange: &buildv1.ImageChangeTrigger{From: &corev1.ObjectReference{Kind: "DockerImage", Name: "mysql", Namespace: "other"}}},
```
Comment (gabemontero) [inline pkg/image/controller/trigger/image_trigger_controller_test.go:646]: yeah this is prevented in the apiserver ... see https://github.com/openshift/openshift-apiserver/blob/4ec57eb921bad25a4e72151137f81cccfd63762e/pkg/build/apis/build/validation/validation.go#L642

**Comment by gabemontero** [comment]
AWS pain/constraints last e2e-aws

**Comment by gabemontero** [comment]
more AWS pain/constraints last e2e-aws

**Comment by gabemontero** [comment]
sure enough ... need to second commit with current apiserver .... investigating today

File: pkg/build/buildutil/util.go [hunk 14]
```diff
@@ -275,3 +276,32 @@ func LabelValue(name string) string {
...
+	return from1.Namespace == from2.Namespace && from1.Name == from2.Name
+}
+
+// GetImageChangeTriggerStatusForImageChangeTrigger returns the ImageChangeTrigger entry in status that corresponds
+// to the supplied ImageChangeTrigger entry in the spec.
+func GetImageChangeTriggerStatusForImageChangeTrigger(triggerSpec *buildv1.ImageChangeTrigger, bc *buildv1.BuildConfig) *buildv1.ImageChangeTriggerStatus {
+	if bc.Status.ImageChangeTriggers == nil || len(bc.Status.ImageChangeTriggers) == 0 {
+		return nil
+	}
+	specFrom := k8stypes.NamespacedName{}
```
Comment (gabemontero) [inline pkg/build/buildutil/util.go:290]: @cdaley - since last time, building `specFrom` did not need to be in the for loop on `bc.Status.ImageChangeTriggers`

File: pkg/build/buildutil/util.go [hunk 15]
```diff
@@ -275,3 +276,32 @@ func LabelValue(name string) string {
...
+	if triggerSpec.From != nil {
+		specFrom.Namespace = triggerSpec.From.Namespace
+		specFrom.Name = triggerSpec.From.Name
+	} else {
+		strategyFrom := buildutil.GetInputReference(bc.Spec.Strategy)
```
Comment (gabemontero) [inline pkg/build/buildutil/util.go:295]: and this else @coreydaley block was missing previously and was why the image registry trigger bc test case was failing ... we were not finding the status entry when the spec entry had a nil from which means it grabs the from of the strategy

File: pkg/image/trigger/buildconfigs/buildconfigs.go [hunk 16]
```diff
@@ -175,8 +176,12 @@ func (r *buildConfigReactor) ImageChanged(obj runtime.Object, tagRetriever trigg
 			newSource = true
 		}
 
-		// LastTriggeredImageID is an image ref, despite the name
-		if latest == p.LastTriggeredImageID {
+		// Determine whether this trigger has fired previously and is recorded in status;
+		// openshift-apiserver updates the status of image change triggers,
+		// including resolving LastTriggeredImageID and LastTriggerTime.
+		pStatus := ocmbuildutil.GetImageChangeTriggerStatusForImageChangeTrigger(p, bc)
+		if pStatus != nil && latest == pStatus.LastTriggeredImageID {
```
Comment (gabemontero) [inline pkg/image/trigger/buildconfigs/buildconfigs.go:183]: this is back to what we want ... where we only need to check our new status last triggered image ID to make sure we do not need to trigger a build

**Review by gabemontero — COMMENTED** [review]
@coreydaley PTAL when you get a sec

**Comment by gabemontero** [comment]
ok e2e-aws passing again ... whew

**Comment by gabemontero** [comment]
@bparees - we have green e2e-aws and e2e-aws-builds and qe / docs / px / lgtm / approve labels ... I believe this non FF item is ready to get the valid-bz item .... wdyt ?

**Comment by gabemontero** [comment]
e2e-aws-upgrade failure is known issue on sippy:


: [sig-network] pods should successfully create sandboxes by other expand_less | 0s
-- | --
1 failures to create the sandbox  ns/openshift-ingress pod/router-default-86f469c969-mpn77 node/ip-10-0-173-128.ec2.internal - never deleted - reason/FailedCreatePodSandBox Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_router-default-86f469c969-mpn77_openshift-ingress_032cab70-8805-4135-99cf-6d5de325db66_0(242bc6680e65a93ad6667f55fe792a55f32c44130f0836693c573616ad7cdc46): Multus: [openshift-ingress/router-default-86f469c969-mpn77]: error getting pod: Get "https://[api-int.ci-op-x10mjccw-3e1a2.origin-ci-int-aws.dev.rhcloud.com]:6443/api/v1/namespaces/openshift-ingress/pods/router-default-86f469c969-mpn77?timeout=1m0s": dial tcp 10.0.157.159:6443: connect: connection refused

https://bugzilla.redhat.com/show_bug.cgi?id=1943334

**Comment by bparees** [comment]
applying valid-bug per no-FF process
